### PR TITLE
fix: Check that the result of pcntl_waitpid is the PID of a managed process

### LINKED ISSUE #1112: `--parallel` option fails if pcntl_waitpid returns non-managed process ID
## Describe the bug

The returned PID of `pcntl_waitpid` is not checked against the list of known PIDs in `src/Runner.php`. In our setup, this call always returns a PID with a value 1 lower than the lowest expected PID in the $childProcs list which causes `phpcs` and `phpcbf` to fail.

E.g. the $childProcs list contains `[5125, 5126]`, but the function first returns `5124` and this causes the runner to crash with the following exception:

```
PHP Fatal error:  Uncaught PHP_CodeSniffer\Exceptions\RuntimeException: Undefined array key 5124 in .../vendor/squizlabs/php_codesniffer/src/Runner.php on line 789 in .../vendor/squizlabs/php_codesniffer/src/Runner.php:624
Stack trace:
#0 .../vendor/squizlabs/php_codesniffer/src/Runner.php(789): PHP_CodeSniffer\Runner->handleErrors()
#1 .../vendor/squizlabs/php_codesniffer/src/Runner.php(560): PHP_CodeSniffer\Runner->processChildProcs()
#2 .../vendor/squizlabs/php_codesniffer/src/Runner.php(216): PHP_CodeSniffer\Runner->run()
#3 .../vendor/squizlabs/php_codesniffer/bin/phpcbf(14): PHP_CodeSniffer\Runner->runPHPCBF()
#4 .../vendor/bin/phpcbf(119): include('...')
#5 {main}
  thrown in .../vendor/squizlabs/php_codesniffer/src/Runner.php on line 624
```

### Code sample
N/A, this bug happens in every project regardless of code snippets.

### To reproduce
1. Create two test PHP files and point phpcbf to them with the following command:  `bash -c 'bash --init-file <(echo "cd /home/vagrant/projects/my-project") -i -t -c "vendor/bin/phpcbf --parallel=8 -p test.php test2.php"'`
2. Notice the exception

## Expected behavior

The parallel option works & phpcs only takes care of managed processes, even when invoked with constructs like these.

## Versions (please complete the following information)

|                                          |                                                                                                                |
|-----------------------------|---------------------------------------------------------------------------- |
| Operating System            | Ubuntu 22.04                                                           |
| PHP version                      | 8.2                                                                                          |
| PHP_CodeSniffer version | 3.13.0                                                                                  |
| Standard                          | PSR12                                                          |
| Install type                       | Composer (local) |

## Additional context

`pcntl_waitpid` already returns the lower value before the first call in pcntl_fork in Runner.php so this is very likely an external factor.

Patching Runner.php with the following code is sufficient to fix the issue, but I am not sure if this is the wanted solution:

```php
        while (count($childProcs) > 0) {
            $pid = pcntl_waitpid(0, $status);
            if ($pid <= 0) {
                continue;
            }
```

change to

```php
        while (count($childProcs) > 0) {
            $pid = pcntl_waitpid(0, $status);
            if ($pid <= 0 || !\array_key_exists($pid, $childProcs)) {
                continue;
            }
```

## Please confirm

- [x] I have searched the issue list and am not opening a duplicate issue.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md) and this is not a [support question](https://github.com/PHPCSStandards/PHP_CodeSniffer/discussions).
- [x] I confirm that this bug is a bug in PHP_CodeSniffer and not in one of the external standards.
- [x] I have verified the issue still exists in the `master` branch of PHP_CodeSniffer.

### PR DESCRIPTION
**Pull request by NanoSector**
# Description
The `--parallel` option spawns processes with `pcntl_fork` and expects the result of `pcntl_waitpid` to be one of its spawned processes. It looks up the temporary output file of a process in the given $childProcs array without checking if the returned process ID is actually managed, which under some circumstances caused by external factors may result in an undefined array index error.

## Suggested changelog entry

N/A, this change likely impacts a very small subset of users.

## Related issues/external references

Fixes #1112 

## Types of changes

- [x] Bug fix _(non-breaking change which fixes an issue)_
- [ ] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_
    - [ ] This change is only breaking for integrators, not for external standards or end-users.
- [ ] Documentation improvement

## PR checklist

- [x] I have checked there is no other PR open for the same change.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md).
- [x] I grant the project the right to include and distribute the code under the BSD-3-Clause license (and I have the right to grant these rights).
- [ ] I have added tests to cover my changes.
  - I looked for existing tests or infrastructure to integrate this change with (with the sample breaking command mentioned in the linked issue), but was unable to find any. I'm open to adding this command as a test in the appropriate place given some pointers.
- [x] I have verified that the code complies with the projects coding standards.
- [x] \[Required for new sniffs\] I have added XML documentation for the sniff.

### CODE REVIEW

File: src/Runner.php [hunk 1]
```diff
@@ -777,7 +777,8 @@ private function processChildProcs($childProcs)
 
         while (count($childProcs) > 0) {
             $pid = pcntl_waitpid(0, $status);
-            if ($pid <= 0) {
+            if ($pid <= 0 || array_key_exists($pid, $childProcs) === false) {
```
Comment (jrfnl) [inline src/Runner.php:780]: Just checking - is there any particular reason to use `array_key_exists()` rather than `isset()` for this change ?

**Review by jrfnl — COMMENTED** [review]
@NanoSector Thank you for this PR and for your willingness to contribute to PHP_CodeSniffer!

While the bug being solved here is very much an edge-case and related to external factors, I also agree that a fix as straight-forward as this one should not be blocked because of that, so happy to merge this.

> I looked for existing tests or infrastructure to integrate this change with (with the sample breaking command mentioned in the linked issue), but was unable to find any. I'm open to adding this command as a test in the appropriate place given some pointers.

This particular issue feels like a perfect example of the type of tests which could be run via [BashUnit](https://github.com/TypedDevs/bashunit) ([Docs](https://bashunit.typeddevs.com/)). Adding a BashUnit test set up to test the CLI commands has been on my "need to look into this" list for a while 😉 

Would you be interested in having a look to see if you can create an initial test setup for PHPCS with BashUnit ?

_Please note: there's no obligation to say "yes" as I also realize that creating an initial BashUnit set up may be more than you bargained for with this PR.
However, I definitely would like to safeguard this fix via automated tests, so if those won't be included in this PR, I'd appreciate it if you could open an issue as a reminder that a test for this particular issue still needs to be added._

**Comment by NanoSector** [comment]
@jrfnl Thanks for your feedback!

BashUnit is something I hadn't heard of before but indeed sounds like a perfect fit to handle cases like these & to application test this package more globally.

I'd definitely be interested in looking into this and it seems straightforward enough, I might be able to take a look later today. This issue isn't blocking for me so I'll append this PR.

File: src/Runner.php (hunk 1 above)
Comment (NanoSector) [reply src/Runner.php:780]: This mostly boils down to habit and preference.

I'm used to working with strictly typed arrays checked by PHPStan so the values in our projects tend to be guaranteed non-nullable; given that I prefer the clearer intention of array_key_exists but your concern is likely over the additional null check isset() performs.

I can either update the PHPDoc to reflect the contents (however the PHPStan level the project uses doesn't really give any guarantees) or I can change it to isset() instead, which would you prefer?

Comment (jrfnl) [reply src/Runner.php:780]: > your concern is likely over the additional null check isset() performs.

Actually, my concern is more one of performance. `isset()` is a language construct, while `array_key_exists()` a function call (and only opcode cached since PHP 7.4).
And yes, while I realize it's a micro-optimization, those type of optimizations actually do have a place in CLI tooling ;-)

> I can either update the PHPDoc to reflect the contents or I can change it to isset() instead, which would you prefer?

If you don't mind, I think I prefer both ;-)

**Comment by NanoSector** [comment]
Looks like something in the patch sequence for the unhappy flow test trips up the tests, disabling that fixes the pipeline - it works on my local machine however. Sorry for the spam you're probably getting.

**Comment by NanoSector** [comment]
@jrfnl I have implemented your feedback & implemented some basic bashunit tests. Few things of note:

1. It uses an in-repository bashunit file, which seemed cleaner than relying on an external installer script.
2. The unhappy flow tests use a temporary file in the current directory; I'd rather make them use a construction with`mktemp` as this would reduce the amount of code needed, but ran into issues where phpcbf would prepend the current working directory to the absolute path so it would fail trying to find the files.
3. The tests for this specific bug are a bit messy... I ran into issues with macOS and its outdated bash version, updating it through Homebrew yielded more process control issues. I opted for a platform-dependent solution and have verified that both commands reproduce the error.
4. I replaced the call to phpcs in the test.yml workflow file, it seemed to me that bashunit makes this obsolete.

Looking forward to your feedback!

File: src/Runner.php [hunk 2]
```diff
@@ -764,7 +764,7 @@ public function processFile($file)
      * The reporting information returned by each child process is merged
      * into the main reporter class.
      *
-     * @param array $childProcs An array of child processes to wait for.
+     * @param array<non-negative-int, non-empty-string> $childProcs An array of child processes to wait for.
```
Comment (jrfnl) [inline src/Runner.php:767]:
```suggestion
     * @param array<int, string> $childProcs An array of child processes to wait for.
```

These annotation should not be tooling specific and `non-negative-int` and `non-empty-string` are PHPStan specific annotations.

**Review by jrfnl — COMMENTED** [review]
@NanoSector Thank you for setting this up, amazing work!

I've looked the commit over. I really like that you've chosen the path of using test scripts. That will help with maintainability and gives contributors the ability to run these tests locally. :+1:

I do have some feedback/questions.

> 1. It uses an in-repository bashunit file, which seemed cleaner than relying on an external installer script.

I'm not so sure about that and I'm not a fan of having code from other projects committed into this repo. The risk of someone making changes to the script and breaking our upgrade path are non-negligable (as this is plain code and not something like a closed PHAR file).

So, I'd rather it be installed in the GH Actions workflow at a specific version and that a section be added to the `CONTRIBUTING` file on what these tests are/when to add tests to the BashUnit setup, and how to install (link to the BashUnit manual + example command) and run the tooling locally.

> 2. The unhappy flow tests use a temporary file in the current directory; I'd rather make them use a construction with `mktemp` as this would reduce the amount of code needed, but ran into issues where phpcbf would prepend the current working directory to the absolute path so it would fail trying to find the files.

Are you talking about the copying of the `Runner.php` file and the patch you apply to that via `tests/EndToEnd/patches/Runner.php_inject_style_error.patch` ?

In my opinion, using real PHPCS files for the tests make the tests quite fragile.
Think: someone submits a PR which updates the `Runner.php` or the `Ruleset.php` file and the patch contains a CS error. Now suddenly the BashUnit tests would also be failing as the output expectations no longer match.
This could be quite confusing for contributors and can lead to time wasted debugging something which doesn't need debugging.

Would it be possible to run these tests on some simple test fixtures instead ? I believe that would also side-step the issue with the current working directory ?

So, instead of having a `patches` subdirectory, there would be a `[Ff]ixtures` subdirectory with some files which can be used in the tests.
Heads-up: You can find some example fixtures like that in the 4.x branch: https://github.com/PHPCSStandards/PHP_CodeSniffer/tree/4.x/tests/Core/Util/ExitCode/Fixtures/ExitCodeTest

And if you then run `phpcbf` in the tests with the `--suffix=.fixed` flag, the original test fixtures would not be changed by the test run (new files with a `.fixed` extension would be created instead and those can be `.gitignore`d for that specific fixtures directory).

That should also remove the need for the `set_up()` method and the `tear_down()` method would then only need to remove the `*.fixed` files from the fixture directory (with the `.gitignore` being a fallback in case the `tear_down()` didn't run for whatever reason).



> 3. The tests for this specific bug are a bit messy... I ran into issues with macOS and its outdated bash version, updating it through Homebrew yielded more process control issues. I opted for a platform-dependent solution and have verified that both commands reproduce the error.

🙌🏻 


> 4. I replaced the call to phpcs in the test.yml workflow file, it seemed to me that bashunit makes this obsolete.

For now, I'd like to keep that test step in place, if you don't mind. I don't mind some QA duplication. Better that, than the tests failing to find an issue before the code gets released.



Other than that, it looks like the tests aren't running/working on the Windows builds ([example](https://github.com/PHPCSStandards/PHP_CodeSniffer/actions/runs/15533994019/job/43728872851#step:21:1)). The Linux jobs all show the testdox output, the Windows jobs don't show any output at all.

I'd expect the step to fail with a non-zero exit code if the tests couldn't run on Windows, preferably with some feedback on why it is failing.
This will probably need some debugging to figure out why the tests aren't running on Windows.
First thing I'd suggest to try, would be adding `shell: bash` to the step and see if that makes a difference.

Then, depending on whether the failure to run on Windows is related to PowerShell or due to an issue with BashUnit itself, you may also want to open an issue in the BashUnit repo about this.

If needs be, I'm okay with skipping these tests on Windows initially and we can look into fixing that at a later point in time.
Making these kind of tests compatible cross-OS will probably not be that straight-forward in a lot of cases, especially for the setup/teardown, so running these tests on Windows may make life a lot more difficult for maintaining these tests, though time will tell.

(Skipping on Windows would mean adding a skip condition to the step in GH Actions + opening an issue about this to be addressed later.)

Let me know what you think.

**Comment by NanoSector** [comment]
@jrfnl Thanks again for your feedback :raised_hands: 

> I'm not so sure about that and I'm not a fan of having code from other projects committed into this repo. The risk of someone making changes to the script and breaking our upgrade path are non-negligable (as this is plain code and not something like a closed PHAR file).

While I understand this, a counter-point here would be bashunit making spontaneous breaking changes to their install script. You can lock the version using this install script but you'd still be relying on external shell code.
If you still prefer this then I'll change it, it's just not necessarily something I agree on.

> Would it be possible to run these tests on some simple test fixtures instead ? I believe that would also side-step the issue with the current working directory ?

Yes, it would, I wasn't aware of the `--suffix` option (I was looking for a `--dry-run` flag instead) :+1: 

> Other than that, it looks like the tests aren't running/working on the Windows builds
> ...
> First thing I'd suggest to try, would be adding shell: bash to the step and see if that makes a difference.

Good catch, I assumed the tests succeeded since the pipeline didn't report errors. I'll give this a try; I don't have a Windows instance to debug on so if this doesn't work I'd prefer disabling the tests for now and opening a separate issue.

**Comment by jrfnl** [comment]
> While I understand this, a counter-point here would be bashunit making spontaneous breaking changes to their install script. You can lock the version using this install script but you'd still be relying on external shell code. If you still prefer this then I'll change it, it's just not necessarily something I agree on.

I get that and I appreciate your thoughtfulness about this, but I'm willing to take that risk.
If the BashUnit team would break their install script at some point in the future, we can re-evaluate, but for now, there is no reason to expect they will (break the script).
And if they would, we report it back and help another open source project to do better (and I expect we wouldn't be the only ones reporting either) 😉 

As an aside, I will be watching releases for the package to keep an eye on things.

> > Would it be possible to run these tests on some simple test fixtures instead ? I believe that would also side-step the issue with the current working directory ?
> 
> Yes, it would, I wasn't aware of the `--suffix` option (I was looking for a `--dry-run` flag instead) 👍

Figured as much. Glad I mentioned it.

> > Other than that, it looks like the tests aren't running/working on the Windows builds
> > ...
> > First thing I'd suggest to try, would be adding shell: bash to the step and see if that makes a difference.
> 
> Good catch, I assumed the tests succeeded since the pipeline didn't report errors. I'll give this a try; I don't have a Windows instance to debug on so if this doesn't work I'd prefer disabling the tests for now and opening a separate issue.

If it helps, I'm on Windows, so we could do a screen share over video if you want me to test some things.

**Comment by jrfnl** [comment]
@NanoSector Just a quick heads-up: the window for getting this merged into the `master` (`3.x`) branch is closing soon (once 4.0.0 has been released, probably next week). Just wanted to make sure you are aware in case it's important for you to have this fix available in 3.x.

**Comment by NanoSector** [comment]
@jrfnl Sorry for the delay in getting back to this. Your feedback should now be implemented.

A separate config file had to be made so that the main config file can exclude the purposefully 'broken' file; this would cause it to report issues otherwise. Please let me know if you have a better solution for this. Pointing phpcs/phpcbf at the file with it listed in an exclusion list causes it to report nothing, likewise including the main phpcs.xml.dist in the end-to-end config excludes these files as well.

The changes I made break on my macOS machine (the process gets put in the background and bashunit stalls) but seem to work on the runners, but I currently lack the time to properly look into this, and the bug originally happened on Linux anyway. I can remove the Darwin check if you'd like, or just make the test succeed on Darwin/macOS.

I also opted to skip the bashunit tests on Windows for now as I don't have the means to properly test this. I have re-added the phpcs invocation in its place. I'll open a ticket for this, however realistically this is not something I'd be interested in implementing.

**Comment by jrfnl** [comment]
@NanoSector Thank you so much for your work on this and my sincere apologies for my delayed response.

The separate config file is a good move as it documents the tests too.

There are some small tweaks I'd still like to make, mostly for consistency in the repo. As it's been a while, I don't want to bother you with it, so I'll push a couple of commits to the PR branch in a minute.

If there are no objections, I'll merge it soon after.

**Comment by jrfnl** [comment]
Closing as manually merged.

**Comment by jrfnl** [comment]
@NanoSector We normally mention people by their name in the changelog. Under what name would you like me to attribute your contribution ? If you prefer not to share your name, I can use your handle instead. Just let me know your preference.

**Comment by NanoSector** [comment]
@jrfnl Happy to see the fix merged 🎉

You can add me as Rick Kerkhof, thank you :) I think it's noteworthy to mention yourself as well for all the research work and fixes!

**Comment by jrfnl** [comment]
> You can add me as Rick Kerkhof, thank you :) I think it's noteworthy to mention yourself as well for all the research work and fixes!

Thanks. Noted. And nah, no need to credit me, I get enough credit as-is ;-)
